binary_search_tree/tree.py

```diff
@@ -17,32 +17,121 @@ def __init__(self):
-        pass
+        if self.root == None:
+            return None
+
+        def find_helper(current):
+            if current.key == key:
+                return current.value
+            elif key < current.key:
+                if current.left is not None:
+                    return find_helper(current.left)
+            else:
+                if current.right is not None:
+                    return find_helper(current.right)
+        return find_helper(self.root)
 
     # Time Complexity: 
     # Space Complexity: 
     def inorder(self):
-        pass
+        return_list = []
+        current = self.root
+        if self.root == None:
+            return return_list
+
+        def in_order_helper(current):
+            if current.left is not None:
+                in_order_helper(current.left)
+            return_list.append({'key': current.key, 'value': current.value})
+            if current.right is not None:
+                in_order_helper(current.right)
+        
+        in_order_helper(current)
+        return return_list
 
     # Time Complexity: 
     # Space Complexity:     
     def preorder(self):
-        pass
+        return_list = []
+        current = self.root
+        if self.root == None:
+            return return_list
+
+        def pre_order_helper(current):
+            return_list.append({'key': current.key, 'value': current.value})
+            if current.left is not None:
+                pre_order_helper(current.left)
+            if current.right is not None:
+                pre_order_helper(current.right)
+        
+        pre_order_helper(current)
+        return return_list
 
     # Time Complexity: 
     # Space Complexity:     
     def postorder(self):
-        pass
+        return_list = []
+        current = self.root
+        if self.root == None:
+            return return_list
+
+        def post_order_helper(current):
+            if current.left is not None:
+                post_order_helper(current.left)
+            if current.right is not None:
+                post_order_helper(current.right)
+            return_list.append({'key': current.key, 'value': current.value})
+        
+        post_order_helper(current)
+        return return_list
 
     # Time Complexity: 
     # Space Complexity:     
     def height(self):
-        pass
+        #check for greatest number
+        #+1 going down stack, -1 going up stack
+        heights = 1
+        max_height = 0
+        current = self.root
+        if self.root == None:
+            return 0
+
+        def height_helper(current):
+            nonlocal heights
+            nonlocal max_height
+            if heights > max_height:
+                max_height = heights
+            if current.left is not None:
+                heights +=1
+                height_helper(current.left)
+            if current.right is not None:
+                heights +=1
+                height_helper(current.right)
+            heights-=1
```

Comment:
This works, but this is a fine example where things get tricky by trying to maintain state that is modified through closure capture. To the point that I think figuring out how to do this without the closure capture would be really helpful to your understanding of recursion. But let me explain first.

To keep track of the current height in `heights`, you have to be very careful to eventually undo every increment with a decrement like on line 131. However, this definitely seems like something that it would be easy to mess up and I had to sit and think about it for a bit.

Like I said, this works, so I don't think I need to go down to Red, but this can be solved without having to maintain state outside of your helper function, so I'm going to mark this Yellow, and consider revisiting this. (You can also do the same for the traversal functions above!)

As a starter hint: let's imagine we're at the root of the tree, and we had an magical "oracle" that will tell us the heights of the left and right subtrees without needing to recurse. Can we figure out the total height from that?

The secret then is that your recursion into the trees will exactly be that oracle function. ;3